### DISCUSSION TITLE
Where to import reflect-metadata?

### DISCUSSION THREAD

**Discussion by matomesc** [vercel/next.js#46805] [Help]
### Summary

I'm trying to use [tsyringe](https://github.com/microsoft/tsyringe) for dependency injection for some services used in my api routes and I have no idea where I'm supposed to `import 'reflect-metadata'`. It's supposed to go in the application entry point but as far as I can tell there is no entry point in nextjs. Any ideas?

### Additional information

```js
Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP Wed Mar 2 00:30:59 UTC 2022
    Binaries:
      Node: 18.12.1
      npm: 8.19.2
      Yarn: 3.3.1
      pnpm: N/A
    Relevant packages:
      next: 13.1.1
      eslint-config-next: 13.1.1
      react: 18.2.0
      react-dom: 18.2.0
```

**Comment by camezcua12** [comment]
I'm currently struggling with the same thing with the new App Router structure. As far as I can see, there's no real entrypoint anymore, or at least it doesn't really show in the docs, as adding `reflect-metadata` to `layout.tsx` doesn't work.

If in your case you're still using the Pages Router (meaning you have a `/pages` or `/src/pages` folder rather than a `/app` one), you should have, or at least be able to create a `_app.tsx` file, where importing `reflect-metadata` would work.

You can read more about it in [here](https://nextjs.org/docs/pages/building-your-application/routing/custom-app).

Still, would be really interested to know how to achieve this but with the new App Router.

**Comment by matomesc** [comment]
@camezcua12 I ended up requiring `reflect-metadata` in `package.json` scripts:

```
{
  ...
  "scripts": {
    "dev": "node --require reflect-metadata ./node_modules/.bin/next dev",
    "build": "next build",
    "start": "node --require reflect-metadata ./node_modules/.bin/next start",
    "lint": "next lint",
  },
  ...
```

**Reply by camezcua12** [reply]
Interesting solution @matomesc, but surely that can't be it. That's more of a workaround or a "hack" than I would consider it a solution. There's something that should be able to be integrated within the codebase to enable reflect-metadata to work.

**Reply by ruohki** [reply]
Since i ran into this exact same issue
you can simply put it into your `NODE_OPTIONS` env variable and it will be executed. No need to modify the package.json other than maybe

```
    "dev": "NODE_OPTIONS=--require reflect-metadata next dev",
    "build": "NODE_OPTIONS=--require reflect-metadata next build",
    "start": "NODE_OPTIONS=--require reflect-metadata next start",
```

Bash
```bash
# Put it into your env
export NODE_OPTIONS='--require reflect-metadata'
```

PowerShell
```powershell
$env:NODE_OPTIONS="--require reflect-metadata"
```

**Reply by aschnebelext** [reply]
Have you managed to make it work on Vercel? I was able to successfully make it work locally with that approach, but not in Vercel.

**Reply by ruohki** [reply]
No it was just for toying around. I was using it to get type-graphql working, but that wont work on vercel sadly

**Reply by wac0013** [reply]
This work for me, minimal changes. 
Just add variable on .env and all commands works. Thanks @matomesc, @ruohki 

```.env
NODE_OPTIONS=--require reflect-metadata 
```

**Comment by ahonn** [comment]
I found that this issue can be solved by modifying the webpack config in next.config.js.

```javascript
// src/polyfills.ts
import 'reflect-metadata';
```

```javascript
/** @type {import('next').NextConfig} */
const nextConfig = {
  webpack: (config) => {
    const originalEntry = config.entry;
    config.entry = async () => {
      const entries = await originalEntry();

      if (
        entries['main.js'] &&
        !entries['main.js'].includes('./src/polyfills.ts')
      ) {
        entries['main.js'].unshift('./src/polyfills.ts'); 
      }

      return entries;
    };

    return config;
  },
};

module.exports = nextConfig;

```

**Reply by aschnebelext** [reply]
Hey @ahonn,

I've tried your approach with a new project (using create-next-app). I installed tsyringe, reflect-metadata, and created the polyfills file importing reflect-metadata.

Finally, I adjusted the next.config.js as you suggested.

I've created a Simple HelloWorld Service, which uses tsyringe and injects the service into the Page directly.

Unfortunately, the problem still exists.

Have you done anything besides what you've mentioned?

**Reply by wouterds** [reply]
Also couldn't get above to work in a new Next.js project.

**Reply by estebangarviso** [reply]
Please, use the solution that I found and let see if it works in your project.

**Reply by wouterds** [reply]
Moved over to Remix, done with Next

**Comment by estebangarviso** [comment]
> ### Summary
> 
> I'm trying to use [tsyringe](https://github.com/microsoft/tsyringe) for dependency injection for some services used in my api routes and I have no idea where I'm supposed to `import 'reflect-metadata'`. It's supposed to go in the application entry point but as far as I can tell there is no entry point in nextjs. Any ideas?
> 
> ### Additional information
> 
> ```js
> Operating System:
> 
>       Platform: linux
> 
>       Arch: x64
> 
>       Version: #1 SMP Wed Mar 2 00:30:59 UTC 2022
> 
>     Binaries:
> 
>       Node: 18.12.1
> 
>       npm: 8.19.2
> 
>       Yarn: 3.3.1
> 
>       pnpm: N/A
> 
>     Relevant packages:
> 
>       next: 13.1.1
> 
>       eslint-config-next: 13.1.1
> 
>       react: 18.2.0
> 
>       react-dom: 18.2.0
> ```
> 
> 
> ### Example
> 
> _No response_

@matomesc
Here is an example with tsyringe
```js
// next.config.js
/** @type {import('next').NextConfig} */
module.exports = {
  experimental: {
    instrumentationHook: true
  }
}
```

```ts
// src/instrumentation.ts
// this is the entrypoint of client and server side
export async function register() {
  // import the the sides that you wish
  await import('./server-container')
  await import('./client-container')
}

// src/client-container
'use client'
import 'reflect-metadata'
import { container } from 'tsyringe'
/*...*/

// src/server-container
import 'reflect-metadata'
import { container } from 'tsyringe'
/*...*/
```

**Reply by caguiclajmg** [reply]
Can't seem to get this to work, it's still throwing the same error (`reflect-metadata` not being imported) on `13.4.12`. What version did you manage to get this to work?

**Reply by estebangarviso** [reply]
Next v13.4.19

**Reply by estebangarviso** [reply]
Here you can create variables in the global scope of the app.

https://stackoverflow.com/questions/38906359/create-a-global-variable-in-typescript

**Reply by thereis** [reply]
Yeah, that helped me to skip the build error. I had to do my internal fixes here but it was a light to guide me through this case. Thanks bro.

**Reply by antoniogiroz** [reply]
Then it's needed to do the same `container.register(...)` in both files, right @estebangarviso?

**Reply by estebangarviso** [reply]
@antoniogiroz depends... if you use DI in the client/server side, just import a container in its respective side.

**Comment by keyiiiii** [comment]
This is how I solved it.
1. import `reflect-metadata` in layout.tsx. (This is for the Server Component)
2. Create `app/ClientLayoutComponent` with Client Component.
3. And also import `reflect-metadata` in `ClientLayoutComponent`


```tsx
// app/layout.tsx
import 'reflect-metadata';
import ClientLayoutComponent from '@/app/ClientLayoutComponent';

export default function RootLayout({
  children
}: {
  children: React.ReactNode;
}) {
  return (
    <html>
      <body>
        <ClientLayoutComponent />
        {children}
      </body>
    </html>
  );
}
```

```ts
// app/ClientLayoutComponent.ts
'use client';

import 'reflect-metadata';

export default function ClientLayoutComponent(): null {
  return null;
}
```

**Reply by olivierr91** [reply]
This does not work either. I am on Next v14.0.4

**Reply by Angeloem** [reply]
This worked for me in Next13. 

On a new project using next14, it doesnt work

**Comment by vs-cristian** [comment]
If you have a `sitemap.ts`(or any file that generates at build time), you should also do the import there.

**Comment by samuelint** [comment]
Here's how I've solved it on `v14.0.4`:
```typescript
//app/reflect-metadata-client-side.ts
'use client';
import 'reflect-metadata';
```
```typescript
//app/layout.tsx
import './reflect-metadata-client-side';
...
```

```typescript
//app/container.ts
import 'reflect-metadata';
import { Container } from 'inversify';
import myModule from '#/lib/myFeature/module';

const container = new Container();
container.load(... myModule);

export default container;
```

Tested on 
- Next `v14.0.4`
- inversify `v6.0.2`

**Reply by jared-sanders** [reply]
This worked as well for me on v14.1.4. 👍

**Reply by olivierr91** [reply]
Worked for me also with tsyringe on v14.1.4

**Comment by chaudhry** [comment]
I encountered a similar issue with Next.js version 14. After some experimentation, I found a workaround by modifying the tsconfig.json file. Here are the changes I made:
```
{
  "compilerOptions": {
    "target": "es5",
    "types": ["reflect-metadata"],    
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true
  }
}
```
By adding these lines to the compilerOptions section, the application was able to run successfully. This solution was tested with Next.js version 14.1.0 and Inversify version 6.0.2.

**Reply by olivierr91** [reply]
I have tried this and still getting the error on random client components.

**Reply by Angeloem** [reply]
> 
> I encountered a similar issue with Next.js version 14. After some experimentation, I found a workaround by modifying the tsconfig.json file. Here are the changes I made:
> ```
> {
>   "compilerOptions": {
>     "target": "es5",
>     "types": ["reflect-metadata"],    
>     "experimentalDecorators": true,
>     "emitDecoratorMetadata": true
>   }
> }
> ```
> By adding these lines to the compilerOptions section, the application was able to run successfully. This solution was tested with Next.js version 14.1.0 and Inversify version 6.0.2.


So you mean you don't import it anywhere in your files?

**Reply by chriskuech** [reply]
This did not work for me either

**Reply by chaudhry** [reply]
Below is my complete tsconfig.json file
```
{
  "compilerOptions": {
    "target": "es5",
    "types": ["reflect-metadata"],    
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "bundler",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "incremental": true,
    "plugins": [
      {
        "name": "next"
      }
    ],
    "paths": {
      "@/*": ["./src/*"]
    }
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"],
  "exclude": ["node_modules"]
}
```

and package.json file

```
{
  "name": "ProjectName",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@fingerprintjs/fingerprintjs": "^3.4.1",
    "@next/third-parties": "^14.1.0",
    "axios": "^1.6.7",
    "inversify": "^6.0.2",
    "next": "^14.1.0",
    "react": "^18",
    "react-dom": "^18",
    "react-hook-form": "^7.52.2",
    "reflect-metadata": "^0.1.13",
    "sharp": "^0.32.6"
  },
  "devDependencies": {
    "@types/node": "^20",
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "autoprefixer": "^10.0.1",
    "eslint": "^8",
    "eslint-config-next": "14.1.0",
    "postcss": "^8",
    "tailwindcss": "^3.3.0",
    "typescript": "^5"
  }
}
```

I hope it will help. If you guys still get error I'll share the boilerplate using Next.js1 4 and inversify.

**Reply by chriskuech** [reply]
I just got it working with https://github.com/vercel/next.js/discussions/46805#discussioncomment-6336139

**Comment by gcmarcello** [comment]
Any updates on this?

**Comment by Et3rnel** [comment]
In Next.js 15.X I had to `import 'reflect-metadata'` on top of my `route.ts` to make it work on the API side.
Putting it in `instrumentation.ts` did not work for me.

**Reply by omerman** [reply]
```
// instrumentation.ts

export async function register() {
  await import('reflect-metadata');
}
```

This worked for me^^

**Comment by wladimirguerra** [comment]
Depending on the code it may be required to polyfill 'reflect-metadata' in the client side. 

I faced this issue. I need to polyfill reflect-metadata in the client side.

[check this answer](https://stackoverflow.com/a/79923033/3776661) and vote if it helped :)

I solved this problem creating this simple component:

```tsx
'use client';
import 'reflect-metadata';
import * as React from 'react';

/**
 * This component is used to polyfill the reflect-metadata package for the client side.
 */
export const ReflectMetadataPolyfill: React.FC = () => null;
```

then i place it in the layout:

```tsx
// app/layout.tsx
import 'reflect-metadata';
import ClientLayoutComponent from '@/app/ClientLayoutComponent';

export default function RootLayout({
  children
}: {
  children: React.ReactNode;
}) {
  return (
    <html>
      <body>
        <ReflectMetadataPolyfill /> {/* <-- */}
        <ClientLayoutComponent />
        {children}
      </body>
    </html>
  );
}
```
